Improve UX of calling abstract-type constructors

Tetralux opened the pull request:
Fixes #20556.

Also hides the `sysimg.jl` method from both the closest-candidates and tab-completion to make things less potentially confusing.

Unless I am missing something, you never intend to call the `sysimg.jl` method directly, if at all, anyway.
And, as can be seen in #20556, it can be an understandable (I think) source of confusion.

Before this PR:

```julia
julia> abstract type X end

julia> X()
ERROR: MethodError: no method matching X()
Closest candidates are:
  X(::Any) where T at sysimg.jl:24

julia> X(x,y) = x+y
X

julia> X()
ERROR: MethodError: no method matching X()
Closest candidates are:
  X(::Any) where T at sysimg.jl:24
  X(::Any, ::Any) at REPL[3]:1

julia> X(<TAB>
(::Type{T})(arg) where T in Base at sysimg.jl:24 X(x, y) in Main at REPL[3]:1
julia> X(
```

After this PR:

```julia
julia> abstract type X end

julia> X()
ERROR: MethodError: no constructors have been defined for X

julia> X(x,y) = x+y
X

julia> X()
ERROR: MethodError: no method matching X()
Closest candidates are:
  X(::Any, ::Any) at REPL[3]:1

julia> X(<TAB>
X(x, y) in Main at REPL[3]:1
julia> X(
```

Tetralux committed:
Make calling abstract type with only sysimg method display a more specific message

Tetralux committed:
Do not show the Base/sysimg.jl method for Type{T} in tab-completion

Tetralux committed:
Hide sysimg.jl method from closest candidates

tkelman commented on test/replutil.jl:
```diff
@@ -499,3 +499,13 @@ foo_9965(x::Int) = 2x
 end
+
+# Issue #20556
+abstract type X end
```
give this type a more specific name

Tetralux committed:
Better type name in tests

nalimilan commented on base/replutil.jl:
```diff
@@ -444,6 +456,10 @@ function show_method_candidates(io::IO, ex::MethodError, kwargs::Vector=Any[])
             buf = IOBuffer()
             tv = Any[]
             sig0 = method.sig
+            # @Hardcoded: don't display the sysimg.jl method
```
What's this? A Twitter account ID? :-)

H-225 replied:
*Sly grin* >)

ararslan replied:
Really though, why the `@`?

H-225 replied:
Reduces programmer friction; the `@` and the capital 'H' make it much more likely you _only_ find what you're actually searching for. (i.e: the tag.)

You could also do things like `#Hardcode: ...`, since this is Julia, but that can blend in with ordinary comments a bit too easily for my liking; the `@` makes it clearer that it is _actually_ a tag.

You can also do things like, `# @Cleanup @Cleanup @Cleanup` etc, if you want it to _really_ stand out when you search for it. >)

nalimilan replied:
This convention isn't used anywhere in the code base currently, so please don't introduce it in this PR. If we really want to use this kind of convention, it needs to be discussed and used by everybody. Else we'll end up with as many conventions as contributors.

pabloferz replied:
What we have used before is TODO or FIX, etc. So if there's need for this, HARCODED would at least be consistent in style.

H-225 replied:
I'll note how few tagged-comments, if any, are anything but `TODO` and `FIX`, which isn't very helpful.

I also occurs to me that doing something like `#TAG` would be better in Julia code, since you can search for that pretty easily, plus, you can do `#[\w_]+:?` to search for _any_ tag.

nalimilan replied:
Again, this it not the place to discuss this. Please open a separate issue if you have a proposal about this.

JeffBezanson commented on test/replutil.jl:
```diff
@@ -499,3 +499,13 @@ foo_9965(x::Int) = 2x
+@test sprint(showerror, (MethodError(AbstractTypeNoConstructors, ()))) == "MethodError: no constructors have been defined for AbstractTypeNoConstructors"
+AbstractTypeNoConstructors(x, y) = x + y
+@test startswith(sprint(showerror, (MethodError(AbstractTypeNoConstructors, ()))), "MethodError: no method matching AbstractTypeNoConstructors()")
+@test !contains(sprint(showerror, (MethodError(AbstractTypeNoConstructors, ()))), "where T at sysimg.jl:24")
```
Could we leave the line number out of the test? It's brittle.

JeffBezanson commented:
LGTM.

Tetralux committed:
Remove line number of sysimg.jl method from tests

Tetralux committed:
Fix tests not checking for qualified names of the type

H-225 commented:
@JeffBezanson 
There are some issues with the `replutil` tests relating to `String(take!(buf)) == color` etc.
I've just checked, and they were present before this PR, AFAICT.

Also, do we want the new message displaying qualified names of things?
I would guess not?

Tetralux committed:
*Actually* add a test for qualified abstract types

tkelman commented on test/replutil.jl:
```diff
@@ -499,3 +499,15 @@ foo_9965(x::Int) = 2x
+module EnclosingModule
+    abstract type AbstractTypeNoConstructors end
+end
+@test sprint(showerror, (MethodError(EnclosingModule.AbstractTypeNoConstructors, ()))) == "MethodError: no constructors have been defined for $(EnclosingModule.AbstractTypeNoConstructors)"
```
could you wrap long lines? we try mostly to stay within about 92 columns

tkelman commented:
I suspect your filtering is changing the printing of something which causes that test to fail

H-225 commented:
I'll check.

Tetralux committed:
Improve tests

- Wrap long lines.
- Make them easier to follow.
- Fix test for tab-completion
  * Should be checking if the method string *starts with* the
    sample, and not `!=` to it.

KristofferC commented:
There's quite a lot of code duplication here. Would it be better to lift it out to a common function? I'm just thinking in case it needs to be updated in the future it might be better to have the logic in one place?

Tetralux committed:
Refactor sysimg method check

KristofferC commented:
- [ ] Run doctests and fixup those that got changed.

Tetralux committed:
Add a comment to the sysimg.jl method for clarity

ararslan commented on base/replutil.jl:
```diff
@@ -666,3 +666,14 @@ function process_backtrace(process_func::Function, t::Vector, limit::Int=typemax
+    if m.module == Base && m.file == Symbol("sysimg.jl") && m.sig == Tuple{Type{T},Any} where T
+        return true
+    end
+    return false
```
This can be simplified to `return m.module == Base && m.file == Symbol("sysimg.jl") && m.sig == Tuple{Type{T},Any} where T`

H-225 replied:
I really shouldn't have written that code when I was tired... >)

Tetralux committed:
Cleanup `is_default_method`

Tetralux committed:
Fix tests; well that was easy...

- since the 'default' method from sysimg.jl is now hidden, there was one less method candidate, and therefore exactly three,
  and therefore, no elipsis anymore.
- removed duplicate test for error message displayed for a MethodError on an abstract type without any constructors
  defined.

Tetralux committed:
Add comments to the tests; make them easier to understand

H-225 commented:
Well... that was easier than I expected. >)

tkelman commented on test/replutil.jl:
```diff
@@ -252,12 +253,6 @@ let
-abstract type T11007 end
-let
-    err_str = @except_str T11007() MethodError
-    @test contains(err_str, "no method matching $(curmod_prefix)T11007()")
```
why is this being deleted?

H-225 replied:
I added a test that covers this case earlier; my test also checks that the message reads 
`no method matching` when there is at least one constructor defined.

Tetralux committed:
Update doctests that fail due to now-hidden sysimg.jl default method

H-225 commented:
Rebased.

H-225 commented:
Doctests have been updated.

Tetralux committed:
Fix rogue syntax error with '&&'

H-225 commented:
RTM?

StefanKarpinski commented:
@KristofferC, if this looks ok to you now, please hit merge.

tkelman commented on test/replutil.jl:
```diff
@@ -20,6 +20,7 @@ method_c1(x::Float64, s::AbstractString...) = true
 buf = IOBuffer()
 Base.show_method_candidates(buf, Base.MethodError(method_c1,(1, 1, "")))
 no_color = "\nClosest candidates are:\n  method_c1(!Matched::Float64, !Matched::AbstractString...)$cfile$c1line"
+@assert length(methods(method_c1)) <= 3
```
why is this an assert instead of a test?

H-225 replied:
Wrote that for debugging a test, but actually, it would be useful to leave this here. Will change it to a test and add a comment.

Tetralux committed:
Use @test instead of @assert

KristofferC commented on base/replutil.jl:
```diff
@@ -444,6 +456,10 @@ function show_method_candidates(io::IO, ex::MethodError, kwargs::Vector=Any[])
             tv = Any[]
             sig0 = method.sig
+            #HARDCODED #REFACTOR: don't display the sysimg.jl method
+            if sig0 == Tuple{Type{T},Any} where T && method.module == Base && method.file == Symbol("sysimg.jl")
```
Could this be `if is_default_method(method)`?

H-225 replied:
Not sure how that slipped through the net. Changing it now... >)

Tetralux committed:
Fix use of old unfactored code

KristofferC commented:
LGTM, anyone feel free to merge when CI passes

H-225 commented:
@KristofferC Appreciated. :)

KristofferC commented:
Thanks for doing this. I really like when it is possible to get rid of bloated noisy output.

H-225 commented:
My sentiment precisely. >)
